File: editor/en/[992]Subarrays with K Different Integers.java

//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
    public int subarraysWithKDistinct(int[] nums, int k) {
        return getMost(nums, k) - getMost(nums, k - 1);
    }

    private int getMost(int[] nums, int k) {
        Map<Integer, Integer> map = new HashMap<>();
        int result = 0;

        for (int left = 0, right = 0; right < nums.length; right++) {
            map.put(nums[right], map.getOrDefault(nums[right], 0) + 1);

            while (map.size() > k) {
                map.put(nums[left], map.get(nums[left]) - 1);

                if (map.get(nums[left]) == 0) {
                    map.remove(nums[left]);
                }

                left++;
            }

            result += right - left + 1;
        }

        return result;
    }
}
//leetcode submit region end(Prohibit modification and deletion)

File: editor/en/[992]Subarrays with K Different Integers.py
# leetcode submit region begin(Prohibit modification and deletion)
from collections import defaultdict


class Solution:
    def subarraysWithKDistinct(self, nums: List[int], k: int) -> int:
        return self.getMost(nums, k) - self.getMost(nums, k - 1)

    def getMost(self, nums: List[int], k: int) -> int:
        count = defaultdict(int)
        left = 0
        result = 0

        for right in range(len(nums)):
            if count[nums[right]] == 0:
                k -= 1
            count[nums[right]] += 1
            while k < 0:
                count[nums[left]] -= 1
                if count[nums[left]] == 0:
                    k += 1
                left += 1
            result += right - left + 1
        return result
# leetcode submit region end(Prohibit modification and deletion)
